app/models/event.rb

```diff
@@ -44,6 +52,36 @@ def formatted_scheduled_time
     scheduled_at.to_formatted_s(:long)
   end
 
+  def is_bookable_by?(applicant)
+    applicant && is_future? && !is_booked_by?(applicant) && !is_participated_by?(applicant) && !is_full?
+  end
+
+  def is_unbookable_by?(applicant)
+    is_future? && is_booked_by?(applicant)
+  end
+
+  def is_booked_by?(applicant)
+    applicant && applicants.exists?(applicant.id)
+  end
+
+  def booked_by!(applicant)
+    Booking.create(event: self, applicant: applicant) if !is_booked_by?(applicant)
+  end
+
+  # TODO: is it worth adding :counter_cache in UserEvent model?
```

Comment:
I tend to leave these kind of performance optimizations until its proven they're needed. Guessing correctly where the bottle-necks are going to be beforehand can be hard. 

However, if you're going to spend a while deciding whether to add it or not, then just add it (it'll save all that time trying to decide).